### DISCUSSION TITLE
Jern - a simple and secure journaling tool

### DISCUSSION THREAD

**Discussion by HoofedEar** [tui-cs/Terminal.Gui#2608] [Show and tell]
Hey I just wanted to share what I made with Terminal.Gui :) it’s a very simple text editor that is meant for journaling, and encrypts  entries securely. I made it as a tool for myself to use on a flash drive that I keep with me. Let me know what you think! 

https://github.com/HoofedEar/Jern

**Comment by tig** [comment]
This is excellent. We should figure out why the keyboard shortcuts are not working on Mac.

**Comment by tznind** [comment]
Nice little program!  Thanks for sharing.  I had a play with it and its cool.

It would be nice if on boot the first time (when there is no .k file), it prompted you to enter a key name.

Prompting (Yes/No/Cancel) to save changes on quit (Alt+Q) would be cool too.

Thanks for referencing back to Terminal.Gui in the about box/readme ❤️.

**Comment by HoofedEar** [comment]
Just curious, is there a Discord? I'd like to start work on adding changes and ideas that people had, but I don't want to spam the Discussions board with questions x) Maybe I can just keep my questions to this main thread?

**Reply by tig** [reply]
Spam away in Discussions or here! Both are fine.

**Comment by HoofedEar** [comment]
Okay I'm encountering an exception that I believe has something to do with how I am switching between Windows

So I have this in my Program.cs file:
```
if (File.Exists("RENAME_ME.k"))
{
    Application.Run<RenameKey>();
}
else
{
    Application.Run<MainWindow>();
}
```
And inside of RenameKey, I do this to switch back to MainWindow:
```
Dispose();
Application.Run<MainWindow>();
```
The error I get is when my application loads RenameKey, then runs MainWindow. Everything seems to work fine, until I try to use a MessageBox.Query. This section throws a NullReferenceException "Object reference not set to an instance of an object." but only after I click "Cool" in the message box. And this only happens when my application runs RenameKey -> MainWindow. If it's just running the MainWindow, it works just fine.
```
var _ = MessageBox.Query(50, 8,
            "About", "Version 1.1\nCreated by HoofedEar\nhttps://hoofedear.itch.io/jern\nPowered by Terminal.Gui",
            "Cool");
```

I tried to display the Rename window as a dialog before, but I ran into an issue where it would leave an artifact in the window if you moved it around and yada yada haha I'm sure I'm doing something funky which is causing this. Any advice on how to better do this? Thank you!

**Reply by BDisp** [reply]
Instead of:
```cs
Dispose();
Application.Run<MainWindow>();
```
do:
```cs
Application.RequestStop();
```

I didn't see the @tznind  answers, sorry for that. You don't need to run again `Application.Run<MainWindow>();` because it was already loaded as @tznind said.

**Comment by tznind** [comment]
> And inside of RenameKey, I do this to switch back to MainWindow

Normally to close a modal window and return to the previous window that was running you should just use:

```
Application.RequestStop();
```

And you don't want to manually dispose of windows.  The framework will automatically call dispose on them for you as long as they were running/added to the closing window.

> I tried to display the Rename window as a dialog before, but I ran into an issue where it would leave an artifact in the window if you moved it around and yada yada haha I'm sure I'm doing something funky which is causing this. Any advice on how to better do this? Thank you!

This sounds like a seperate issue. If you are able to create a repro we can probably advise on what is causing artifacts after close.  Or fix it if it is a bug.

**Reply by HoofedEar** [reply]
Ah that worked! Thank you! haha I just assumed that it would kill the entire application with that statement, but good to know!

**Reply by dodexahedron** [reply]
As a tip for grokkability, you could always alias Application, while in the context of anything that isn't actually the "application" from your PoV.

Calling the alias something other than application, whether you do it with a local variable reference or with a using, is a totally valid way of helping yourself out with things like that in General. 🙂

As long as it's the right reference and is appropriately scoped, you won't run into any GC issues or anything, and TG won't even know you did it.

Or, you can do a `using static Terminal.Gui.Application;` and then call it without the Application qualifier, as if it were locally scoped in that file (so just `Run<T>()` instead of `Application.Run<T>()` etc for static members).

**Comment by HoofedEar** [comment]
Not sure if this is already known, but I noticed when running my application on macOS 14.5, after upgrading it to .NET 8 and version 1.17, the top and bottom parts of the window look weirdly cut off:
![image](https://github.com/gui-cs/Terminal.Gui/assets/1261392/f346e674-2c8f-44f2-9463-74ceb4f27dea)
I'm spinning up a Windows 11 VM right now but last I looked, I don't this this issue is present on Windows

**Reply by HoofedEar** [reply]
Here's how it looks on Windows, notice how the top and bottom section goes all the away across
![Screenshot 2024-06-17 at 3 46 45 PM](https://github.com/gui-cs/Terminal.Gui/assets/1261392/e5f625e4-ced0-44dd-862f-81848c5b8f65)

**Reply by BDisp** [reply]
I think this is related to the type of font used or the terminal used.
I can only test on macOS 12 Monterey because I can no longer install a newer one on a VM and it doesn't have this behavior.

**Reply by HoofedEar** [reply]
Interesting, looks like you're right about the terminal used making a difference. Left is the default Terminal, right is iTerm2. I tried adjusting the font in iTerm2 but that didn't seem to make a difference. Weird! 
![Screenshot 2024-06-17 at 4 02 49 PM](https://github.com/gui-cs/Terminal.Gui/assets/1261392/15cc5b38-dec2-4cd3-87c1-a067ba78f9ac)

**Reply by tznind** [reply]
I have also encountered this, bits of screen getting 'cut out' (i.e. not rendered).  IT seems to happen primarily with 'empty space' or HLine.  If you find a solution I would be really interested.

It wasn't this way initially, its like either terminal or mac OS update broke it.

**Reply by dodexahedron** [reply]
Same.

I get it seemingly at random in tmux and screen sometimes, too.

Happens outside of them, too, if I've run some other program in the session that may have been badly behaved or if I piped some output from something that might not have been plain text.

I don't _think_ there's anything we can really do about it, without a full terminal reset.

**Reply by BDisp** [reply]
I'm convinced that is a terminal related.

**Reply by dodexahedron** [reply]
I concur completely.

While we may have _something_, _somewhere_ that exacerbates it, I don't think we are ever the first place that the broken behavior starts.

**Reply by BDisp** [reply]
It seems to me that it is not respecting a fixed-size font if it is not processing simple Ansi characters, handling lines (-) as spaces that do not count towards cursor advancement.

**Reply by HoofedEar** [reply]
I'm sure it's been noted already but this seems to be fixed in the latest v1 version, so thanks for that! :)

**Comment by HoofedEar** [comment]
Would now be a good time to transition to v2? 
I did some work today to attempt migrating (using the latest prerelease version listed: 2.0.0-prealpha.1829)  
And so far it was straight forward! The only thing I'm having trouble with is moving to the new StatusBar/Shortcut and Commands stuff.

Here is the old code:
![image](https://github.com/user-attachments/assets/a88b2a49-64d4-4561-8ab5-7bfffaa83b36)

Here is the new code:
![image](https://github.com/user-attachments/assets/6ad55db9-dc1e-4f7a-b93e-bc160baeb4e9)

However I noticed that when I press Esc, it doesn't call my SaveBeforeQuit method. So I realized I needed to do something different using Keybindings?
Basically, I'm confused about how to remap command keys (I noticed that I could do something like `AddCommand`) and how I can use an async method with it? I noticed it requires a nullable bool but does true mean success and false mean failed? (I found this https://gui-cs.github.io/Terminal.GuiV2Docs/api/Terminal.Gui.View.AddCommand.html in the docs but couldn't find anything about that)

I like the new changes using keybindings, just need some guidance on how how to adapt here haha thanks!

**Reply by dodexahedron** [reply]
V2 is still alpha and the API surface is not stable yet.

Definitely not the time to switch to it for production applications or anything that has significant dependency on TG yet.

For new development, I guess it's up to you and your comfort level with the prospect of any and all parts of the public API surface not having any guarantee of being the same from one moment to the next, depending on what happens as we continue development.

The public API surface should be largely stable once we declare beta, with the likelihood of breaking changes being relatively low, at that point. Until then, though, alpha is alpha. 😅

**Reply by tig** [reply]
I disagree. 

V2 is very API stable. The only remaining places where there will be significant churn is around TopLevel. 

We need more devs targeting it. 

For true production apps, sure. But we should be encouraging devs to party. 

For this particular question I'll take a deeper look tomorrow and answer the question then.

**Reply by HoofedEar** [reply]
Hey thanks for the clarification! Haven't had a chance to get back to looking at this but I'm hoping to take another look at it tomorrow 👍

**Reply by HoofedEar** [reply]
Ah figured it out. Instead of having to use `Key.AltMask | Key.S`, I needed to use `Key.S.WithAlt`, but now I can use `WithCtrl` and the shortcut actually works as intended on MacOS so that's nice!

**Reply by HoofedEar** [reply]
Is there a way to programmatically set specific settings? I basically just want to force dialogs to center their buttons, globally in the application. I'd like to avoid using an external .config file if at all possible

**Reply by tig** [reply]
Yes, you can set, for example, `Dialog.DefaultButtonAlignment` which is `static`. You'll need to do it after `Init`.

I recommend, though, that you use a .config file and put in your apps `/Resources`. See how UI Catalog does this. 

This way, users of your app can override your settings if they want.

Note that the final look & feel for v2 is not settled and we'd love your feedback. If all devs start centering dialog buttons because we choose the default wrong that would be a shame. See 

- #3493

And feel free to voice your opinions there!

**Reply by HoofedEar** [reply]
Thanks so much tig! 

I'll be sure to post there once I tweak things some more and learn about what look I end up going with. So far just centering the buttons for Dialogs is the only thing I've come across haha
